docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,332 @@
+```kotlin
+val client = S3Client.fromEnvironment {
+    region = "us-west-2"
+}
+client.addCompressionAlgorithm(...)
+```
+
+### The Trait
+Below is an abbreviated example of the requestCompression [trait](https://smithy.io/2.0/spec/behavior-traits.html?highlight=requestcompression#requestcompression-trait) in use:
+
+```smithy
+@requestCompression(
+    encodings: ["gzip", "custom"]
+)
+operation ...
+```
+
+The SDK has its own list of supported compression algorithms as a list of `CompressionAlgorithm`s in the client config.
+
+```kotlin
+private val compressionAlgorithms: List<CompressionAlgorithm> = listOf(
+    GzipCompressionAlgorithm()
+)
+```
+
+### Behavior
+Each algorithm will be attempted in the order they appear in the trait until one that is supported by the runtime is found. If no algorithm is found the request will continue without compression applied.
+
+After compression occurs, the `Content-Encoding` header will contain the name of the compression algorithm used. It will be appended to the list of content encodings if the header was already present with some value.
+
+```http request
+// There was no encoding in header before compression
+POST /upload-data HTTP/1.1
+Host: example.com
+Content-Type: application/json
+Content-Encoding: gzip
+
+// There was an encoding in header before compression
+POST /upload-data HTTP/1.1
+Host: example.com
+Content-Type: application/json
+Content-Encoding: br, gzip 
+```
+
+### Compression Algorithms
+Given that Kotlin is multiplatform there should be multiple implementations of each compression algorithm. As of the time of writing this doc (11/07/2023) gzip is the only supported compression algorithm by the server side and the only one supported for
+the SDK. For the time being, only gzip for JVM is supported specifically. This is done by using the following packages from the Java standard library: `GZIPInputStream` and `GZIPOutputStream`. The other KMP targets are not implemented, and are planned to be implemented using the CRT.
+
+New algorithms can be added as they start becoming available/required using the compression algorithm interface and by adding them to the list of supported algorithms. The same goes for non JVM KMP targets, these can be added by filling in the `actual` declarations for the target.
+
+#### Interfaces
+This interface is used to represent a compression algorithm.
+
+```kotlin
+/**
+ * Represents a compression algorithm to be used for compressing request payloads on qualifying operations
+ */
+public interface CompressionAlgorithm {
+    /**
+     * The ID of the compression algorithm
+     */
+    public val id: String
+    
+    /**
+     * Compresses a payload
+     */
+    public suspend fun compress(stream: ByteStream?): ByteStream
+}
+```
+
+The `id` is used to match a supported `CompressionAlgorithm` to a requested compression algorithm from the trait. The `compress` function is used to compress HTTP bodies.
```

Comment:
> The `compress` function is used to compress HTTP bodies

I think the input to this function should be an `HttpBody` then, instead of a `ByteStream`?